docs: add modules documentation pages

shahednasser opened the pull request:
- Added modules overview page
- Added Create Module guide. This is a general guide that explains the process (similar to plugins) which will be referred to in other guides like Create a Cache Module guide.

Closes DOCS-676, DOCS-677, DOCS-678

shahednasser commented on docs/content/development/modules/create.mdx:
```diff
@@ -0,0 +1,321 @@
+
+---
+
+### Step 5: Export Module
```
Not super sure about the details here

carlos-r-l-rodrigues commented on docs/content/development/modules/create.mdx:
```diff
@@ -0,0 +1,321 @@
+Where:
+
+- `service`: This is the only required property to be exported. It should be the main service your module exposes.
+- `loaders`: (optional) an array of functions used to perform an action while loading the module. For example, you can log a message that the module has been loaded.
```
As an example of a `loader` we can mention "Establishing the Database connection"

carlos-r-l-rodrigues commented on docs/content/development/modules/create.mdx:
```diff
@@ -0,0 +1,321 @@
+
+- `service`: This is the only required property to be exported. It should be the main service your module exposes.
+- `loaders`: (optional) an array of functions used to perform an action while loading the module. For example, you can log a message that the module has been loaded.
+- `migrations`: (optional) an array of paths to migration files that should run when the `migration` command is used with Medusa's CLI.
```
```suggestion
- `migrations`: (optional) an array of objects containing database migrations that should run when the `migration` command is used with Medusa's CLI.
```

carlos-r-l-rodrigues commented on docs/content/development/modules/create.mdx:
```diff
@@ -0,0 +1,321 @@
+- `loaders`: (optional) an array of functions used to perform an action while loading the module. For example, you can log a message that the module has been loaded.
+- `migrations`: (optional) an array of paths to migration files that should run when the `migration` command is used with Medusa's CLI.
+- `models`: (optional) an array of entities that your module creates.
+- `runMigrations`: (optional) a function that can be used to define migrations to run when the `migration run` command is used with Medusa's CLI. The migrations will only run if they haven't already.
```
```suggestion
- `runMigrations`: (optional) a function that can be used to define migrations to run when the `migration run` command is used with Medusa's CLI. The migrations will only run if they haven't already. This will only be executed if the module's scope is `isolated`.
```

carlos-r-l-rodrigues commented on docs/content/development/modules/create.mdx:
```diff
@@ -0,0 +1,321 @@
+- `migrations`: (optional) an array of paths to migration files that should run when the `migration` command is used with Medusa's CLI.
+- `models`: (optional) an array of entities that your module creates.
+- `runMigrations`: (optional) a function that can be used to define migrations to run when the `migration run` command is used with Medusa's CLI. The migrations will only run if they haven't already.
+- `revertMigration`: (optional) a function can be used to define how migrations should be reverted when the `migration revert` command is used with Medusa's CLI.
```
```suggestion
- `revertMigration`: (optional) a function can be used to define how migrations should be reverted when the `migration revert` command is used with Medusa's CLI. This will only be executed if the module's scope is `isolated`.
```

carlos-r-l-rodrigues commented on docs/content/development/modules/create.mdx:
```diff
@@ -0,0 +1,321 @@
+
+Where:
+
+- `service`: This is the only required property to be exported. It should be the main service your module exposes.
```
```suggestion
- `service`: This is the only required property to be exported. It should be the main service your module exposes. The exposed service must implement all the declared methods on that particular module's Interface located at `@medusajs/types`. eg: [ICacheService](https://github.com/medusajs/medusa/blob/4e9d257d3bf76703ef5be8ca054cc9f0f7339def/packages/types/src/cache/service.ts) interface)
```

carlos-r-l-rodrigues commented on docs/content/development/modules/create.mdx:
````diff
@@ -0,0 +1,321 @@
+```json title=tsconfig.json
+{
+  "compilerOptions": {
+    "lib": [
+      "es5",
+      "es6",
+      "es2019"
+    ],
+    "target": "es5",
````
```suggestion
    "lib": ["es2020"],
    "target": "2020",
```

carlos-r-l-rodrigues commented on docs/content/development/modules/create.mdx:
```diff
@@ -0,0 +1,321 @@
+    "noImplicitThis": true,
+    "allowJs": true,
+    "skipLibCheck": true,
+    "downlevelIteration": true
```
we can remove "downlevelIteration": true, and target es6+ for new modules.

carlos-r-l-rodrigues commented on docs/content/development/modules/create.mdx:
```diff
@@ -0,0 +1,321 @@
+    "prepare": "cross-env NODE_ENV=production npm run build",
+    "build": "tsc --build",
+  },
+  "peerDependencies": {
+    "@medusajs/medusa": "^1.7.13"
+  },
```
remove the peerDependencies

carlos-r-l-rodrigues commented on docs/content/development/modules/create.mdx:
```diff
@@ -0,0 +1,321 @@
+
+In this document, you’ll learn how to create a module, test the module, publish it to NPM, and install it in a Medusa backend.
+
+This document covers the general steps of creating the module, but does not explore actually implementing any functionality in it.
```
### 
```suggestion
This document covers the general steps of creating an internal module, but does not explore actually implementing any functionality in it.
```

carlos-r-l-rodrigues commented on docs/content/development/modules/create.mdx:
```diff
@@ -0,0 +1,321 @@
+- `moduleType` is the type of your module. For example, if your module is a cache module, it should be changed to `cacheService`.
+- `<module_name>` is the name of your module as defined in the `package.json` of your module.
+- `options` should hold any options of your module, if necessary.
+
```
Maybe we could expand this and add the `resources` property that is inside `moduleType`. at the same level as the `resolve`

resouces can be:

`shared` (this is the default) - All the Awilix container is shared with the module. Including the database connection. The db connection can be skipped if the module is sharing resources with Medusa's Core.

`isolated` - The module will receive an empty Awilix container, and only dependencies will be registered there. These dependencies are defined at the Module's definition and cannot be changed. The module will be responsible for managing it's own database connection, and other resources.

carlos-r-l-rodrigues reviewed (commented):
Overall it looks great. At this moment we only support "internal" modules so I believe it is OK to keep this documentation as it is.
When we support external modules I think we'll need to split it and be explicity when configuring a module as internal or external.

shahednasser reviewed (commented):
Thanks a lot for your review Carlos! Your comments were very helpful. I made changes based on your comments.

carlos-r-l-rodrigues commented on docs/content/development/modules/create.mdx:
```diff
@@ -0,0 +1,344 @@
+
+In this document, you’ll learn how to create a module, test the module, publish it to NPM, and install it in a Medusa backend.
+
+This document covers the general steps of creating an internal module, but does not explore actually implementing any functionality in it. An internal module is a module that extends or replaces a functionality within the Medusa backend, such as the cache or event bus functionality.
```
It is correct. Maybe we could be very specific just to avoid any future confusion with external module. Something like this:

An internal module is a javascript/typescript module that is loaded by Medusa core as part of the  application to extend or replace a functionality within the Medusa backend, such as the cache or event bus functionality.

olivermrbl replied:
+1

olivermrbl commented on docs/content/development/modules/create.mdx:
```diff
@@ -0,0 +1,344 @@
+
+---
+
+## Step 7: Publish Module
```
comment: Optional

shahednasser commented on docs/content/development/modules/create.mdx:
```diff
@@ -0,0 +1,344 @@
+
+---
+
+## Step 1: Create an NPM Project
```
@olivermrbl can you give me an example of how it would work with file reference? Would help with the writing 😇

carlos-r-l-rodrigues replied:
@shahednasser to load modules declared in the project the `medusa-config` will be the same, but the "resolve" property will point to a local file or a folder with an index.{ts,js}. The file just has to be exported following the `ModuleExports` type.

```js
module.exports = {
  // ...
  modules: { 
    // ...
    inventory: {
      resolve: "../project_modules/custom_inventory", 
      // resolve: "../custom_modules/inventory.js",
      options: {
        // options if necessary
      },
      // optional
      resources: "shared",
    },
  },
}
```

carlos-r-l-rodrigues replied:
Correct. That is it.

olivermrbl replied:
+1 – this is nice!

shahednasser commented:
@olivermrbl / @carlos-r-l-rodrigues I changed up the create documentation to focus on creating the documentation without NPM, and added a separate publish documentation explaining the steps to publish the module

olivermrbl reviewed (approved):
LGTM - I like the split 💯

carlos-r-l-rodrigues reviewed (approved):
It's great
